--- a/drivers/spi/spi-mt65xx.c
+++ b/drivers/spi/spi-mt65xx.c
<<<<<<< HEAD
@@ -1228,8 +1228,15 @@ static int mtk_spi_probe(struct platform
=======
@@ -1237,8 +1237,15 @@ static int mtk_spi_probe(struct platform
>>>>>>> 2ab15ff3e3c42f513cfc49ab3057b7eaa0392a78
 	if (ret < 0)
 		return dev_err_probe(dev, ret, "failed to enable hclk\n");
 
+	ret = clk_prepare_enable(mdata->sel_clk);
+	if (ret < 0) {
+		clk_disable_unprepare(mdata->spi_hclk);
+		return dev_err_probe(dev, ret, "failed to enable sel_clk\n");
+	}
+
 	ret = clk_prepare_enable(mdata->spi_clk);
 	if (ret < 0) {
+		clk_disable_unprepare(mdata->sel_clk);
 		clk_disable_unprepare(mdata->spi_hclk);
 		return dev_err_probe(dev, ret, "failed to enable spi_clk\n");
 	}
